audio/OpenAL/OpenALAudioProvider.cpp

```diff
@@ -0,0 +1,245 @@
+            {
+                auto provider = reinterpret_cast<OpenALAudioProvider*>(ptr);
+                provider->LogOpenALMessages(l, m);
+            }), this);
+        }
+
+
+        //Device and Context Init
+        _device = alcOpenDevice(nullptr);
+        if(!_device)
+        {
+            std::string error = GetALError();
+            throw Exceptions::InitialisationFailureException(
+                "OpenAL failed to create an audio device!", error);
+        }
+        _context = alcCreateContext(_device, nullptr);
+        if(!_context)
+        {
+            std::string error = GetALError();
+            throw Exceptions::InitialisationFailureException(
+                "OpenAL failed to create and attach a proper context!", error);
+        }
+        alcMakeContextCurrent(_context);
+    }
+
+    OpenALAudioProvider::~OpenALAudioProvider()
+    {
+        Dispose();
+    }
+
+    void OpenALAudioProvider::Dispose()
+    {
+        alGetError();
+        alSourceStopv(static_cast<int>(_sources.size()), reinterpret_cast<ALuint*>(_sources.data()));
+        GetALError();
+        alDeleteSources(static_cast<int>(_sources.size()), reinterpret_cast<ALuint*>(_sources.data()));
+        GetALError();
+        _sources.clear();
+        alDeleteBuffers(static_cast<int>(_buffers.size()), reinterpret_cast<ALuint*>(_buffers.data()));
+        GetALError();
+        _buffers.clear();
+        alcMakeContextCurrent(NULL);
+        alcDestroyContext(_context);
+        alcCloseDevice(_device);
+    }
+
+    uint32_t OpenALAudioProvider::OpenSource(AudioSourceContext& context)
+    {
+        uint32_t source = 0;
+        alGetError();
+        alGenSources(1, &source);
+        GetALError();
+        alSourcef(source, AL_GAIN, context.Volume);
+        GetALError();
+        alSourcef(source, AL_PITCH, context.Pitch);
+        GetALError();
+        alSourcei(source, AL_LOOPING, static_cast<int>(context.Loop));
+        GetALError();
+        _sources.emplace_back(static_cast<uint32_t>(source));
+        return source;
+    }
+
+    // void OpenALAudioProvider::CloseSource()
+    // {
+    //     alDeleteSources(1, &_outputSource);
+    // }
+
+    void OpenALAudioProvider::PlaySource(uint32_t sourceId)
+    {
+        alGetError();
+        alSourcePlay(sourceId);
+        GetALError();
+    }
+
+    void OpenALAudioProvider::StopSource(uint32_t sourceId)
+    {
+        alGetError();
+        alSourceStop(sourceId);
+        GetALError();
+    }
+
+    void OpenALAudioProvider::PauseSource(uint32_t sourceId)
+    {
+        alGetError();
+        alSourcePause(sourceId);
+        GetALError();
+    }
+
+    std::string OpenALAudioProvider::GetALError()
+    {
+        auto err = alGetError();
+        switch (err)
+        {
+            case AL_INVALID_NAME:
+            {
+                _logger->error("A bad ID or name was passed to the OpenAL function.");
```

Comment:
Are we directly using spdlog due to the deprecation of the old APIs?

Reply:
Yes - in this case its easier to do so, and since there's some uncertainty around how the APIs will be used going forward (with the great big refactor) it was just cleaner to do so, so that I could continue to keep the static lib isolated to itself